### PR TITLE
Add JVM IR target in build_snapshot_train mode

### PR DESCRIPTION
**Pull request by ALikhachev**

### CODE REVIEW

**Comment by udalov** [comment]
As in https://github.com/ktorio/ktor/pull/2175#issuecomment-721058147, I suggest not to commit API dumps for JVM IR, but rather mute the tests if possible (with the option to unmute locally).

File: benchmark/build.gradle [hunk 1]
```diff
@@ -8,32 +8,56 @@ apply plugin: "me.champeau.gradle.jmh"
...
-    compile 'com.google.code.gson:gson:2.8.1'
-    compile 'com.esotericsoftware:kryo:4.0.0'
+    implementation 'com.squareup.okio:okio:1.13.0'
+    implementation 'com.google.code.gson:gson:2.8.1'
+    implementation 'com.esotericsoftware:kryo:4.0.0'
 
-    compile project(':kotlinx-serialization-core')
-    compile project(':kotlinx-serialization-json')
-    compile project(':kotlinx-serialization-protobuf')
+    jvmImplementation project(':kotlinx-serialization-core')
```
Comment (sandwwraith) [inline benchmark/build.gradle:55]: Why there is a need for separate configuration? I though benchmarks can consume the same class/jar files

File: build.gradle [hunk 2]
```diff
@@ -163,11 +170,27 @@ subprojects {
 
     apply from: rootProject.file('gradle/teamcity.gradle')
     // Configure publishing for some artifacts
-    if (project.name != "benchmark" && project.name != "guide") {
+    if (project.name != "benchmark" && project.name != "guide" && !(project.name in multiplatformProjects)) {
```
Comment (sandwwraith) [inline build.gradle:173]: Why make this condition and move publishing setup to other gradle files?

File: formats/hocon/build.gradle [hunk 3]
```diff
@@ -14,15 +14,17 @@
  *  limitations under the License.
  */
 
-apply plugin: 'kotlin'
+apply plugin: 'kotlin-multiplatform'
```
Comment (sandwwraith) [inline formats/hocon/build.gradle:17]: Will it affect publishing? I'd rather try to avoid making this subproject multiplatform

Comment (sandwwraith) [reply formats/hocon/build.gradle:17]: Also, it would be nice to set up source dirs so actual files should not be moved

**Review by sandwwraith — COMMENTED** [review]
Can you briefly describe how changes will affect build process? As far as I can see, there will be new `compileJvmIr` and `jvmIrTest` tasks? Why the new dependency configuration is needed? Will jvm ir artifacts be published and how?

File: build.gradle (hunk 2 above)
Comment (ALikhachev) [reply build.gradle:173]: There is some problem at the moment: user-defined variant attributes (JVM compiler type in our case) are not taken into account when publishing is set up before applying kotlin-multiplatform plugin

File: benchmark/build.gradle (hunk 1 above)
Comment (ALikhachev) [reply benchmark/build.gradle:55]: Adding of JVM IR target causes ambiguity in dependency resolution, so we need to mark dependency either with JVM or JVM IR

**Comment by ALikhachev** [comment]
The main purpose of those changes is to ensure that JVM IR compiler is correct. We compile JVM sources using JVM IR compiler and run test classes against it's output the same as with the old compiler. JVM IR jar doesn't pretend to be published.

File: formats/hocon/build.gradle (hunk 3 above)
Comment (ALikhachev) [reply formats/hocon/build.gradle:17]: It shouldn't negatively affect publishing

File: benchmark/build.gradle [hunk 4]
```diff
@@ -8,32 +8,56 @@ apply plugin: "me.champeau.gradle.jmh"
...
+    jvmImplementation project(':kotlinx-serialization-json')
+    jvmImplementation project(':kotlinx-serialization-protobuf')
+    jvmImplementation project(':kotlinx-serialization-json')
```
Comment (hfhbd) [inline benchmark/build.gradle:58]: Why did you implement `-json` twice?

Comment (ALikhachev) [reply benchmark/build.gradle]: This dependency was duplicated after rebase, and I didn't notice it. Thank you!

File: benchmark/build.gradle (hunk 1 above)
Comment (sandwwraith) [reply benchmark/build.gradle:55]: Can you please add the comment explaining this in details to configuration delcaration?

File: formats/hocon/build.gradle (hunk 3 above)
Comment (sandwwraith) [reply formats/hocon/build.gradle:17]: Well, there's now `kotlinx-serialization-hocon` and `kotlinx-serialization-hocon-jvm` artifacts. The first one depends on the second, so users shouldn't face issues unless they're using maven: #1143

**Review by sandwwraith — APPROVED** [review]

**Comment by sandwwraith** [comment]
Note: due to hocon setup change, this PR also seems to fix #1189

File: benchmark/build.gradle (hunk 1 above)
Comment (ALikhachev) [reply benchmark/build.gradle:55]: We create configuration named 'jvmImplementation' [here](https://github.com/Kotlin/kotlinx.serialization/blob/7d39fd395249e314f842b3e17caa1ec6b8167a68/benchmark/build.gradle#L13), set JVM compiler attribute with value **jvm** and then declare dependencies to that configuration. Then we configure **jmhTask** to use this configuration. Do you mean add the comment to the code?

**Comment by ALikhachev** [comment]
During private discussion with @qwwdfsad we've decided to try another way. I'll open new PR.

**Pull request closed without merging**